Provenance v1: add verification section, clarify model

MarkLodato opened the pull request:
This explains how to verify provenance without expanding the requirements of the SLSA Build levels. As discussed at the SLSA Specification SIG meeting on 2023-02-13, adding verification to the existing levels seems like it adds too much complexity. This PR tries to strike the balance of still describing the concept without making it a SLSA requirement.

Also clarify the model:

- Diagram:
  - Simplify the diagram's colors (orange = "to be verified", green = "trusted", remove red).
  - Mark the build cache as green since it's covered by the SLSA Build requirement, thus trusted.
  - Add a legend (#593.)
  - Use the same text color as the website.
- Better explain what `buildType` means.
- Clarify communication with caches and the platform.

Preview: https://deploy-preview-618--slsa.netlify.app/provenance/v1

joshuagl commented on docs/provenance/v1.md:
```diff
@@ -481,9 +483,152 @@ The timestamp of when the build completed.
+    "G" or "H".)
+-   [Threat "H"]: Verification by the consumer covers compromise of the package
+    in transit. (Many ecosystems already address this threat using package
+    signatures.)
```
Is it worth adding:
```suggestion
    signatures or checksums.)
```
?

In many open source ecosystems signatures are not particularly common, but checksums/hash-checking are increasingly common. Is it worth mentioning those here?

joshuagl commented on docs/provenance/v1.md:
```diff
@@ -481,9 +483,152 @@ The timestamp of when the build completed.
+threats further up the supply chain.
+
+One possible approach is to recursively [verify][Verification] each entry in
+`resolvedDependencies`. A [Verification Summary Attestation (VSA)][VSA] can be
```
```suggestion
`resolvedDependencies`. A [Verification Summary Attestation (VSA)][VSA] can
```

joshuagl commented on docs/provenance/v1.md:
```diff
@@ -481,9 +483,152 @@ The timestamp of when the build completed.
+
+Up front:
+
+-   Establish roots of trust, meaning the recognized pairs of builder public key
```
Nit: Should we _identify_, rather than _establish_, roots of trust?

MarkLodato replied:
What about *configure*?

joshuagl replied:
WFM!

joshuagl commented on docs/provenance/v1.md:
```diff
@@ -481,9 +483,152 @@ The timestamp of when the build completed.
+-   Establish roots of trust, meaning the recognized pairs of builder public key
+    identity + `builder.id`, mapped to SLSA Build levels. Different
+    organizations might use different roots of trust, but usually a verifier
+    uses the same root of trust for all packages.
```
It might be useful to provide a pedagogical example of what this might look like, in the form of a JSON snippet perhaps?

MarkLodato replied:
FYI, my previous commit said "max level" but I realize now it's just "level". There are no futher checks that could cause it to be downgraded.

joshuagl commented on docs/provenance/v1.md:
```diff
@@ -481,9 +483,152 @@ The timestamp of when the build completed.
+or prevent this adversary from injecting unofficial behavior into the package.
+Example threats in this category include building from an unofficial fork or
+abusing a build parameter to modify the build. Usually expectations identify the
+canonical source repository (which is the main external parameter) and any other
```
nit: I think this a bit more explicit/easier to understand if we use the field name:

```suggestion
canonical source repository (which is the main entry in `externalParameters`) and any other
```

joshuagl commented on docs/provenance/v1.md:
```diff
@@ -481,9 +483,152 @@ The timestamp of when the build completed.
+| `buildType` | To ensure that `externalParameters` are interpreted as intended |
+| `externalParameters` | To prevent an adversary from injecting unofficial behavior |
+
+Verifiers SHOULD reject unrecognized external parameters. It is acceptable to
```
```suggestion
Verifiers SHOULD reject unrecognized `externalParameters`. It is acceptable to
```

joshuagl commented on docs/provenance/v1.md:
```diff
@@ -481,9 +483,152 @@ The timestamp of when the build completed.
+that any value in the range is safe. Implementations need not special-case the
+`buildType` if JSON comparisons are sufficient.
+
+Possible models (not exhaustive):
```
This could do with some more context to help it flow, otherwise the jump from verifying expectations to setting expectations is a bit jarring.
```suggestion
Possible models for implementing expectation setting in package ecosystems (not exhaustive):
```

joshuagl reviewed (approved):
Thanks, this is great. Including a diagram with the steps mapped to threats is a nice touch.

MarkLodato commented:
Thanks, I addressed your suggestions. In some cases I added a little text, though the roots of trust bullet I rewrote substantially. I'm not sure about the example so please take a look to see if it's good.

joshuagl reviewed (approved):
Thanks, this LGTM. The example does a good of demonstrating the concept. 👍

MarkLodato commented:
I'll merge this now since we're getting close to the RC deadline, but as always nothing is set in stone. Feel free to open further. issues or comments.
